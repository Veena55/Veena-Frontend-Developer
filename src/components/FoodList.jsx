import React, { useEffect, useState } from 'react'
import FoodCard from './FoodCard'
import Filter from './Filter'
import Sort from './Sort'
import Dropdown from './Dropdown'
import { useDispatch, useSelector } from 'react-redux'
import { errorSelector, fetchFoodItems, filteredByAreaData, foodSelector, loadingSelector } from '../redux/slices/FoodItemsSlice'
import { CgSpinner } from 'react-icons/cg'
import axios from 'axios'

const FoodList = () => {
    const [isDropdownVisible, setIsDropdownVisible] = useState(false);
    const [areaList, setAreaList] = useState([]);
    const [area, setArea] = useState('India');

    const dispatch = useDispatch();

    //foodItemsList
    const foodItemsList = useSelector(foodSelector);

    //loading state
    const isLoading = useSelector(loadingSelector);

    //error state
    const error = useSelector(errorSelector);


    useEffect(() => {
        console.log("hi");
        dispatch(fetchFoodItems());
        console.log(foodItemsList);
    }, [dispatch])


    const handleFilter = () => {
        console.log("hi");
        setIsDropdownVisible(prev => !prev)
    }
    const getAllArea = async () => {
        const { data } = await axios.get("https://www.themealdb.com/api/json/v1/1/list.php?a=list");
        setAreaList(data.meals);
    }

    useEffect(() => {
        getAllArea();
    }, [])

    const getListOfFoodByArea = (e, areaToFilter) => {
        console.log(e.target.value);
        if (e.target.value == "on") {
            setArea(areaToFilter);
            dispatch(filteredByAreaData(areaToFilter));
            setIsDropdownVisible(false);
        }
    }


    if (isLoading) {
        return <div className="flex justify-center"><CgSpinner className='text-themeColor animate-spin duration-200' size={50} /></div>
    }

    if (error) {
        return <div className='text-red-500 font-bold text-xl italic capitalize'>Error.. {error} !</div>
    }

    return (
        <div>
            <div>
                <h1 className="font-bold text-2xl">Restaurants with online food delivery in {area}</h1>
            </div>
            <div className="py-5 flex gap-3">
                <div className="relative">
                    <Filter handleFilter={handleFilter} />
                    {isDropdownVisible && <Dropdown areas={areaList} getListOfFoodByArea={getListOfFoodByArea} />}
                </div>
                <Sort />
            </div>
            <div className="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-10 py-3">
                {console.log(foodItemsList)}
                {foodItemsList.map((item, index) => {
                    console.log(item);
                    return <FoodCard key={item.idMeal} item={item} />

                })}
                {/* <FoodCard />
                <FoodCard />
                <FoodCard /> */}
            </div>
        </div>
    )
}

export default FoodList