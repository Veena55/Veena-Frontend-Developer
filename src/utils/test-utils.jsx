import React from "react"
import { render } from "@testing-library/react"
import { setupStore } from "../store"
import { Provider } from "react-redux"

export function renderWithProviders(ui, extendedRenderOptions = {}) {
    const {
        preloadedState = {
            foodItems: [],
            isLoading: false,
            foodItemById: {},
            error: null,
            sortDirection: true, // true -> accending, false -> descending
            isModalOpen: false, // handle modal open & close(by default close)
            idToFilter: null,
            pageNo: 0
        },
        // Automatically create a store instance if no store was passed in
        store = setupStore(preloadedState),
        ...renderOptions
    } = extendedRenderOptions

    const Wrapper = ({ children }) => (
        <Provider store={store}>{children}</Provider>
    )

    // Return an object with the store and all of RTL's query functions
    return {
        store,
        ...render(ui, { wrapper: Wrapper, ...renderOptions })
    }
}