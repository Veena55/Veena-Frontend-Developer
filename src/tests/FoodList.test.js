import React from 'react';
import FoodList from '../components/FoodList';
import { renderWithProviders } from '../utils/test-utils';

describe('<FoodCard>', () => {
  it('should render component', () => {
    expect(renderWithProviders(<FoodList />)).toMatchSnapshot();
  });



});