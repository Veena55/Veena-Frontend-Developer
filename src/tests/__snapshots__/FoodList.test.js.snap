// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FoodCard> should render component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex justify-center"
      >
        <svg
          class="text-themeColor animate-spin duration-200"
          fill="none"
          height="50"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="50"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M12 19C15.866 19 19 15.866 19 12C19 8.13401 15.866 5 12 5C8.13401 5 5 8.13401 5 12C5 15.866 8.13401 19 12 19ZM12 22C17.5228 22 22 17.5228 22 12C22 6.47715 17.5228 2 12 2C6.47715 2 2 6.47715 2 12C2 17.5228 6.47715 22 12 22Z"
            fill="currentColor"
            fill-rule="evenodd"
            opacity="0.2"
          />
          <path
            d="M2 12C2 6.47715 6.47715 2 12 2V5C8.13401 5 5 8.13401 5 12H2Z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex justify-center"
    >
      <svg
        class="text-themeColor animate-spin duration-200"
        fill="none"
        height="50"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="50"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M12 19C15.866 19 19 15.866 19 12C19 8.13401 15.866 5 12 5C8.13401 5 5 8.13401 5 12C5 15.866 8.13401 19 12 19ZM12 22C17.5228 22 22 17.5228 22 12C22 6.47715 17.5228 2 12 2C6.47715 2 2 6.47715 2 12C2 17.5228 6.47715 22 12 22Z"
          fill="currentColor"
          fill-rule="evenodd"
          opacity="0.2"
        />
        <path
          d="M2 12C2 6.47715 6.47715 2 12 2V5C8.13401 5 5 8.13401 5 12H2Z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
